pkg/storage/cacher.go

```diff
@@ -389,7 +399,7 @@ func newCacherListerWatcher(storage Interface, resourcePrefix string, newListFun
 // Implements cache.ListerWatcher interface.
 func (lw *cacherListerWatcher) List() (runtime.Object, error) {
 	list := lw.newListFunc()
-	if err := lw.storage.List(context.TODO(), lw.resourcePrefix, 0, Everything, list); err != nil {
+	if err := lw.storage.List(context.TODO(), lw.resourcePrefix, "", Everything, list); err != nil {
```

Comment:
Oh, right. Of course. :)

On Mon, Dec 7, 2015 at 11:32 PM, Wojciech Tyczynski <
notifications@github.com> wrote:

> In pkg/storage/cacher.go
> <https://github.com/kubernetes/kubernetes/pull/18207#discussion_r46922348>
> :
>
> > @@ -389,7 +399,7 @@ func newCacherListerWatcher(storage Interface, resourcePrefix string, newListFun
> >  // Implements cache.ListerWatcher interface.
> >  func (lw *cacherListerWatcher) List() (runtime.Object, error) {
> >  	list := lw.newListFunc()
> > -	if err := lw.storage.List(context.TODO(), lw.resourcePrefix, 0, Everything, list); err != nil {
> > +	if err := lw.storage.List(context.TODO(), lw.resourcePrefix, "", Everything, list); err != nil {
>
> Actually no.
> The "cacherListerWatcher" is the class that is reponsible for updating the
> internal cacher based on etcd state. So what it is doing it has an open
> "watch" for all resources of a given type to etcd and all "real watchers"
> are just accessing cache.
>
> The List() here is used to either initiate the first watch or to relist in
> case e.g. falling out of etcd window.
> This means that there will never be a corresponding user request here.
>
> Why do we change it to ""? It's because we would like to force the list
> from etcd - if you pass an empty string, this will cause exactly that.
> So this is WAI.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/18207/files#r46922348>.
>